### PR TITLE
Migrate CPU Backend to allocate memory at runtime.

### PR DESCRIPTION
**Pull request by gcatron**
*Description*: Migrate CPU Backend to allocate and initialize memory at runtime. Added a new runtimeInfo object that contains information needed for runtime memory allocation. 
*Testing*: ninja test passes, AOT bundles still work, testing with run.sh --iterations=20 -cpu -time showed minimal performance difference.
*Documentation*: N/A
Progress on #1904

### CODE REVIEW

File: include/glow/Backends/CompiledFunction.h [hunk 1]
```diff
@@ -16,11 +16,25 @@
 #ifndef GLOW_BACKENDS_COMPILEDFUNCTION_H
 #define GLOW_BACKENDS_COMPILEDFUNCTION_H
 
+#include <glow/Graph/Nodes.h>
```
Comment (nadavrot) [inline include/glow/Backends/CompiledFunction.h]: Why do we need nodes.h?  Can we use a forward declaration?

File: include/glow/Backends/CompiledFunction.h [hunk 2]
```diff
@@ -16,11 +16,25 @@
...
 #include <unordered_map>
 
 namespace glow {
 
 class Context;
+class Placeholder;
+
+struct runtimeSymbolInfo {
```
Comment (nadavrot) [inline include/glow/Backends/CompiledFunction.h]: Doxygen please.

File: include/glow/Backends/CompiledFunction.h [hunk 3]
```diff
@@ -16,11 +16,25 @@
...
+  size_t size;
+  size_t offset;
+};
+struct runtimeInfo {
```
Comment (nadavrot) [inline include/glow/Backends/CompiledFunction.h]: doxygen for the class decl and all members.

File: include/glow/Backends/CompiledFunction.h (hunk 2 above)
Comment (nadavrot) [reply include/glow/Backends/CompiledFunction.h]: We use the "_" suffix for class members unless they are really simple pods like (point.x  point.y).

File: lib/Backends/CPU/AllocationsInfo.cpp [hunk 4]
```diff
@@ -47,36 +46,16 @@ void AllocationsInfo::allocateWeightVars(const IRFunction *F,
...
-  } else {
-    // Allocate based on size as reported by the formal type of Placeholders
-    for (auto &v : F->getGraph()->getParent()->getPlaceholders()) {
-      assert(isa<WeightVar>(F->getWeightForNode(v)));
-      auto *w = cast<WeightVar>(F->getWeightForNode(v));
-      auto numBytes = w->getSizeInBytes();
-      size_t addr = mutableWeightVarsAllocator.allocate(numBytes, w);
-      allocatedAddressed_[w] = addr;
-    }
+  // Allocate based on size as reported by the formal type of Placeholders
```
Comment (nadavrot) [inline lib/Backends/CPU/AllocationsInfo.cpp]: Period at the end of the sentence.

File: lib/Backends/CPU/AllocationsInfo.cpp [hunk 5]
```diff
@@ -47,36 +46,16 @@ void AllocationsInfo::allocateWeightVars(const IRFunction *F,
...
+  for (auto &v : F->getGraph()->getParent()->getPlaceholders()) {
+    assert(isa<WeightVar>(F->getWeightForNode(v)));
```
Comment (nadavrot) [inline lib/Backends/CPU/AllocationsInfo.cpp]: Could you add a comment here that explains what's going on?

File: lib/Backends/CPU/AllocationsInfo.cpp [hunk 6]
```diff
@@ -101,6 +80,43 @@ void AllocationsInfo::allocateWeightVars(const IRFunction *F,
   });
 }
 
+void AllocationsInfo::collectConstants(const IRFunction *F) {
+
+  // At compile time condense constants to a single block of memory
+  // This allows the graph to go away after compile time
```
Comment (nadavrot) [inline lib/Backends/CPU/AllocationsInfo.cpp]: period.

File: lib/Backends/CPU/AllocationsInfo.cpp [hunk 7]
```diff
@@ -101,6 +80,43 @@ void AllocationsInfo::allocateWeightVars(const IRFunction *F,
...
+  baseConstantWeightVarsStore_ =
+      (uint8_t *)alignedAlloc(constantWeightVarsMemSize_, TensorAlignment);
+  for (auto &v : F->getGraph()->getParent()->getConstants()) {
+    assert(isa<WeightVar>(F->getWeightForNode(v)));
+    auto *w = cast<WeightVar>(F->getWeightForNode(v));
+    auto payload = v->getPayload().getUnsafePtr();
+    auto numBytes = w->getSizeInBytes();
+    auto addr = allocatedAddressed_[w];
+    // copy weight to offset.
```
Comment (nadavrot) [inline lib/Backends/CPU/AllocationsInfo.cpp]: Capital letter to start the sentence.

File: lib/Backends/CPU/AllocationsInfo.cpp [hunk 8]
```diff
@@ -101,6 +80,43 @@ void AllocationsInfo::allocateWeightVars(const IRFunction *F,
...
+    memcpy(baseConstantWeightVarsStore_ + addr, payload, numBytes);
+  }
+}
+
+std::unique_ptr<runtimeInfo>
+AllocationsInfo::generateRuntimeInfo(const IRFunction *F) {
+  std::unique_ptr<runtimeInfo> info(new runtimeInfo{});
```
Comment (nadavrot) [inline lib/Backends/CPU/AllocationsInfo.cpp]: Can you add a few comments here?

File: lib/Backends/CPU/AllocationsInfo.h [hunk 9]
```diff
@@ -16,6 +16,7 @@
 #ifndef GLOW_BACKENDS_CPU_ALLOCATIONSINFO_H
 #define GLOW_BACKENDS_CPU_ALLOCATIONSINFO_H
 
+#include "glow/Backends/CompiledFunction.h"
```
Comment (nadavrot) [inline lib/Backends/CPU/AllocationsInfo.h]: Do we need this header or can we use a forward declaration?

File: lib/Backends/CPU/CPUBackend.cpp [hunk 10]
```diff
@@ -41,16 +41,16 @@ namespace {
...
 
-/// Emit the entry point for JIT called "jitmain". It simply calls the main
-/// entry of the module with the constant concrete addresses of all the memory
-/// areas. Since these addresses are constants, the LLVM optimizer will constant
-/// propagate them into relative addressing computations and the like and
-/// produce a very efficient code that uses absolute addressing whenever
-/// possible.
+/// Emit the entry point for JIT called "jitmain".
+/// Function has the following API:
+/// void jitmain(uint8_t *baseConstantWeightVars, uint8_t *baseInoutWeightVars,
```
Comment (nadavrot) [inline lib/Backends/CPU/CPUBackend.cpp]: Can you add two spaces between the comment marker and the "void" to make sure that the code looks like doc and not some commented out code? I am not sure if clang-format will allow this.

File: lib/Backends/CPU/CPUFunction.cpp [hunk 11]
```diff
@@ -13,28 +13,87 @@
...
+CPUFunction::CPUFunction(std::unique_ptr<llvm::orc::GlowJIT> JIT,
+                         std::unique_ptr<runtimeInfo> runtimeInfo)
+    : JIT_(std::move(JIT)), runtimeInfo_(std::move(runtimeInfo)) {}
+
+CPUFunction::~CPUFunction() { free(runtimeInfo_->constants); }
+
+void CPUFunction::loadFunctionOnDevice() {
+  baseConstantWeightVarsAddress_ = (uint8_t *)alignedAlloc(
+      runtimeInfo_->constantWeightVarsMemSize, TensorAlignment);
+  // Copy Constants to device
```
Comment (nadavrot) [inline lib/Backends/CPU/CPUFunction.cpp]: Period.

File: lib/Backends/CPU/CPUFunction.h [hunk 12]
```diff
@@ -27,12 +27,29 @@ class CPUFunction final : public CompiledFunction {
   /// The LLVM JIT engine. The jit must be initialized after the ctor
   /// initializes the LLVM backends.
   std::unique_ptr<llvm::orc::GlowJIT> JIT_;
-  /// This represents the heap, that stores the activations at runtime.
-  void *heap_;
+
+  /// runtimeInfo contains symbol offsets and allocation sizes
```
Comment (nadavrot) [inline lib/Backends/CPU/CPUFunction.h]: Period.

File: lib/ExecutionEngine/ExecutionEngine.cpp [hunk 13]
```diff
@@ -76,6 +76,7 @@ void glow::updateInputPlaceholdersByName(Context &ctx, Module *mod,
 
 void ExecutionEngine::run(Context &ctx) {
   assert(function_ && "No function has been compiled");
+  // TODO call setup functions here not in execute.
```
Comment (nadavrot) [inline lib/ExecutionEngine/ExecutionEngine.cpp]: Can you elaborate here?

For example: "We are working on migrating our API from A to B. In the process we will implement C. See task #123456 "

File: lib/Backends/CPU/CPUFunction.cpp [hunk 14]
```diff
@@ -13,28 +13,87 @@
...
     JitFuncType funcPtr = reinterpret_cast<JitFuncType>(address.get());
-    funcPtr();
+    funcPtr(baseConstantWeightVarsAddress_, baseMutableWeightVarsAddress_,
+            baseActivationsAddress_);
+    fetchFromDevice(ctx);
   } else {
     GLOW_ASSERT(false && "Error getting address.");
   }
-}
+  freeAllocations();
```
Comment (nadavrot) [inline lib/Backends/CPU/CPUFunction.cpp]: new line at the end of the file please.

File: include/glow/Backends/CompiledFunction.h (hunk 2 above)
Comment (opti-mix) [reply include/glow/Backends/CompiledFunction.h]: Type names in our sources typically start with an upper-case letter.

Please provide comments for the type and all of its fields/methods.

File: include/glow/Backends/CompiledFunction.h [hunk 15]
```diff
@@ -16,11 +16,25 @@
...
+  std::unordered_map<std::string, runtimeSymbolInfo> symbolTable;
+  uint8_t *constants;
+  size_t constantWeightVarsMemSize;
```
Comment (opti-mix) [inline include/glow/Backends/CompiledFunction.h]: Are all these `MemSize` mutable? If not, may be they should be `const`?

File: include/glow/Backends/CompiledFunction.h (hunk 2 above)
Comment (opti-mix) [reply include/glow/Backends/CompiledFunction.h]: I'm also not quite happy about the name of the type. It sounds too general to me. May be use something more specific? And/or may be it should be in its own namespace, e.g. `glow::runtime` or `glow::execution`?

File: lib/Backends/CPU/AllocationsInfo.h [hunk 16]
```diff
@@ -75,6 +72,11 @@ struct AllocationsInfo {
   /// Number all allocations and weight variables by assigning them unique
   /// numbers.
   void numberValues(const IRFunction *F);
+
+  /// Collect Constants into a single block of memory
```
Comment (opti-mix) [inline lib/Backends/CPU/AllocationsInfo.h]: Add period at the end of this sentence.

File: lib/Backends/CPU/AllocationsInfo.h [hunk 17]
```diff
@@ -75,6 +72,11 @@ struct AllocationsInfo {
...
+  void collectConstants(const IRFunction *F);
+
+  std::unique_ptr<runtimeInfo> generateRuntimeInfo(const IRFunction *F);
```
Comment (opti-mix) [inline lib/Backends/CPU/AllocationsInfo.h]: Add a doxygen comment.

File: lib/Backends/CPU/CPUFunction.cpp [hunk 18]
```diff
@@ -13,28 +13,87 @@
...
+        runtimeInfo_->symbolTable.find(std::string(PH.first->getName()));
+    assert(symbolInfo != runtimeInfo_->symbolTable.end() && "Symbol not found");
+    auto addr = symbolInfo->second.offset;
+    auto numBytes = symbolInfo->second.size;
+    // copy PH to allocated memory.
+    memcpy(baseMutableWeightVarsAddress_ + addr, payload, numBytes);
+  }
+}
+
+void CPUFunction::fetchFromDevice(Context &ctx) {
```
Comment (opti-mix) [inline lib/Backends/CPU/CPUFunction.cpp]: Since you have a `loadMutablesOnDevice` method, the name for this one should be more symmetric, e.g. `fetchMutablesFromDevice`.

But in other backends, we usualy use names like 'copySomethingToDevice` and `copySomethingFromDevice`.

File: lib/Backends/CPU/CPUFunction.h [hunk 19]
```diff
@@ -27,12 +27,29 @@ class CPUFunction final : public CompiledFunction {
...
+  uint8_t *baseMutableWeightVarsAddress_;
 
 public:
   /// Ctor.
-  CPUFunction(std::unique_ptr<llvm::orc::GlowJIT> JIT, void *heap);
+  CPUFunction(std::unique_ptr<llvm::orc::GlowJIT> JIT,
+              std::unique_ptr<runtimeInfo> runtimeInfo);
+
+  /// Copy Function and Constant Memory to device.
+  void loadFunctionOnDevice();
```
Comment (opti-mix) [inline lib/Backends/CPU/CPUFunction.h]: I wonder if we always want to combine copying of a function (on a CPU no code needs to be copied, but on accelerators we may need to actually copy code) and copying of constant weights into a single function like `loadFunctionOnDevice`? May be it should be two functions? Or a single function that internally calls two different functions?

Comment (opti-mix) [reply lib/Backends/CPU/CPUFunction.h]: A more general question: Should the CompiledFunction class be responsible for copying to/from devices and for memory allocation? It seems to me that these are runtime/execution aspects and thus belong into a separate API. CompiledFunction objects themselves are passive objects that should be manipulated/executed by other subsystems IMHO.

File: lib/Backends/CPU/CPUFunction.cpp [hunk 20]
```diff
@@ -13,28 +13,87 @@
...
   assert(sym && "Unable to JIT the code!");
-  using JitFuncType = void (*)(void);
+  using JitFuncType =
+      void (*)(uint8_t * constantWeightVars, uint8_t * mutableWeightVars,
+               uint8_t * activations);
   auto address = sym.getAddress();
   if (address) {
     JitFuncType funcPtr = reinterpret_cast<JitFuncType>(address.get());
-    funcPtr();
+    funcPtr(baseConstantWeightVarsAddress_, baseMutableWeightVarsAddress_,
```
Comment (opti-mix) [inline lib/Backends/CPU/CPUFunction.cpp]: This part related to the invocation of a JITed function looks good!

File: lib/Backends/CPU/CPUBackend.cpp [hunk 21]
```diff
@@ -41,16 +41,16 @@ namespace {
...
+/// uint8_t *baseActivations);
 static void emitJitMain(LLVMIRGen &irgen) {
   AllocationsInfo &allocationsInfo = irgen.getAllocationsInfo();
   llvm::Type *voidTy = llvm::Type::getVoidTy(irgen.getLLVMContext());
-  llvm::FunctionType *jitFuncTy = llvm::FunctionType::get(voidTy, {}, false);
+  auto int8PtrTy = llvm::Type::getInt8PtrTy(irgen.getLLVMContext());
```
Comment (opti-mix) [inline lib/Backends/CPU/CPUBackend.cpp]: The changed logic for creation of the `jitmain` function looks good.

File: lib/Backends/CPU/CPUFunction.h (hunk 19 above)
Comment (opti-mix) [reply lib/Backends/CPU/CPUFunction.h]: > Compiled function already has execute which is responsible for the actual execution, so I can see a case for keeping all runtime logic co-located. But maybe it's cleaner to separate that functionality to a separate API. Can we remove execute too? Then CompiledFunction would be just the code itself.

Yeah, I think mid term we most likely want the CompiledFunction to be just the code itself and have a separate API/logical system components to run/execute one or more CompiledFunctions. In case of a CPU execution, such a component may be e.g. responsible for managing threads, deciding which CompiledFunction should run in which thread, handle thread local memory @rdzhabarov mentioned, etc.

File: include/glow/Backends/CompiledFunction.h (hunk 1 above)
Comment (gcatron) [reply include/glow/Backends/CompiledFunction.h]: I may be going about it wrong, but it seems it needs to be there as part of the hash validity check on the unordered_map.

File: include/glow/Backends/CompiledFunction.h (hunk 2 above)
Comment (gcatron) [reply include/glow/Backends/CompiledFunction.h]: I like the glow::runtime namespace, I moved it in there and renamed it to RuntimeBundle and RuntimeSymbolInfo.

File: lib/Backends/CPU/AllocationsInfo.cpp [hunk 22]
```diff
@@ -47,36 +46,16 @@ void AllocationsInfo::allocateWeightVars(const IRFunction *F,
...
-      size_t addr = mutableWeightVarsAllocator.allocate(numBytes, w);
-      allocatedAddressed_[w] = addr;
-    }
+  // Allocate based on size as reported by the formal type of Placeholders
+  for (auto &v : F->getGraph()->getParent()->getPlaceholders()) {
+    assert(isa<WeightVar>(F->getWeightForNode(v)));
+    auto *w = cast<WeightVar>(F->getWeightForNode(v));
+    auto numBytes = w->getSizeInBytes();
+    size_t addr = mutableWeightVarsAllocator.allocate(numBytes, w);
+    allocatedAddressed_[w] = addr;
```
Comment (rdzhabarov) [inline lib/Backends/CPU/AllocationsInfo.cpp]: nit: allocatedAddressed_ - typo here?

File: lib/Backends/CPU/AllocationsInfo.cpp [hunk 23]
```diff
@@ -101,6 +80,43 @@ void AllocationsInfo::allocateWeightVars(const IRFunction *F,
...
+    auto *w = cast<WeightVar>(F->getWeightForNode(v));
+    auto payload = v->getPayload().getUnsafePtr();
+    auto numBytes = w->getSizeInBytes();
+    auto addr = allocatedAddressed_[w];
+    // copy weight to offset.
+    memcpy(baseConstantWeightVarsStore_ + addr, payload, numBytes);
+  }
+}
+
+std::unique_ptr<runtimeInfo>
```
[the remark this answers is not in the document]
Comment (gcatron) [reply lib/Backends/CPU/AllocationsInfo.cpp]: Good catch, the thought was it would be better not to copy around a large object but it's actually quite small. Now passing directly.

File: include/glow/Backends/CompiledFunction.h [hunk 24]
```diff
@@ -16,12 +16,44 @@
...
+#include "glow/Graph/Nodes.h"
 #include <unordered_map>
 
 namespace glow {
 
 class Context;
+namespace runtime {
+/// Runtime Symbol Info
+/// Contains Information for initialization and handling of symbol at runtime.
+/// \p size is the size in bytes of the symbol.
```
Comment (opti-mix) [inline include/glow/Backends/CompiledFunction.h]: The comments for members are put right before the members.

File: include/glow/Backends/CompiledFunction.h [hunk 25]
```diff
@@ -16,12 +16,44 @@
...
+struct RuntimeSymbolInfo {
+  size_t size;
+  // namespace runtime
+  size_t offset;
+};
 
+/// Runtime Bundle
+/// Contains the information needed to be passed forward from compile time to
+/// runtime.
+/// \p symbolTable is a map from symbol name to a RuntimeSymbolObject.
```
Comment (opti-mix) [inline include/glow/Backends/CompiledFunction.h]: Same here.

File: lib/Backends/CPU/AllocationsInfo.cpp [hunk 26]
```diff
@@ -94,13 +75,50 @@ void AllocationsInfo::allocateWeightVars(const IRFunction *F,
...
+    memcpy(baseConstantWeightVarsStore_ + addr, payload, numBytes);
+  }
+}
+
+runtime::RuntimeBundle
+AllocationsInfo::generateRuntimeBundle(const IRFunction *F) {
+  runtime::RuntimeBundle info(constantWeightVarsMemSize_,
+                              mutableWeightVarsMemSize_, activationsMemSize_);
+  std::unordered_map<std::string, runtime::RuntimeSymbolInfo> symbolTable;
+  // info->mutableWeightVarsMemSize = mutableWeightVarsMemSize_;
```
Comment (opti-mix) [inline lib/Backends/CPU/AllocationsInfo.cpp]: Remove the commented out code?

File: lib/Backends/CPU/AllocationsInfo.cpp [hunk 27]
```diff
@@ -47,44 +47,25 @@ void AllocationsInfo::allocateWeightVars(const IRFunction *F,
...
-      assert(isa<WeightVar>(F->getWeightForNode(v)));
-      auto *w = cast<WeightVar>(F->getWeightForNode(v));
-      auto numBytes = w->getSizeInBytes();
-      size_t addr = mutableWeightVarsAllocator.allocate(numBytes, w);
-      allocatedAddressed_[w] = addr;
-    }
+  // Compute the offsets and total memory requirements for Placeholders.
+  for (auto &v : F->getGraph()->getParent()->getPlaceholders()) {
+    // Get the WeightVar for each Placeholder to calculate offsets.
+    assert(isa<WeightVar>(F->getWeightForNode(v)));
```
Comment (opti-mix) [inline lib/Backends/CPU/AllocationsInfo.cpp]: asserts typically should have an error message.

File: lib/Backends/CPU/CPUFunction.h [hunk 28]
```diff
@@ -27,12 +27,31 @@ class CPUFunction final : public CompiledFunction {
   /// The LLVM JIT engine. The jit must be initialized after the ctor
   /// initializes the LLVM backends.
   std::unique_ptr<llvm::orc::GlowJIT> JIT_;
-  /// This represents the heap, that stores the activations at runtime.
-  void *heap_;
+
+  /// runtimeBundle contains symbol offsets and allocation sizes.
+  runtime::RuntimeBundle runtimeBundle_;
+
+  uint8_t *baseConstantWeightVarsAddress_;
```
Comment (opti-mix) [inline lib/Backends/CPU/CPUFunction.h]: Add doxygen comments for all of  these base addresses.

File: lib/Backends/CPU/CPUFunction.h [hunk 29]
```diff
@@ -27,12 +27,31 @@ class CPUFunction final : public CompiledFunction {
...
+  CPUFunction(std::unique_ptr<llvm::orc::GlowJIT> JIT,
+              runtime::RuntimeBundle runtimeBundle);
+
+  /// Copy Function to device, an empty function for CPU.
+  void copyFunctionToDevice(){};
+  /// Copy Constants to device.
+  void copyConstantsToDevice();
+  /// Allocate Mutable buffers on device, Activations and Placeholders.
+  void allocateMutableBuffersOnDevice();
+  /// Copy Mutable data to device
```
Comment (opti-mix) [inline lib/Backends/CPU/CPUFunction.h]: Dot at the end of each comment.

File: lib/Backends/CPU/AllocationsInfo.cpp (hunk 23 above)
[the remark this answers is not in the document]
Comment (rdzhabarov) [reply lib/Backends/CPU/AllocationsInfo.cpp]: For return value it does not really matter because of RVO.
For passing it around you could use const reference.

File: include/glow/Backends/CompiledFunction.h [hunk 30]
```diff
@@ -25,27 +25,30 @@ class Context;
 namespace runtime {
 /// Runtime Symbol Info
 /// Contains Information for initialization and handling of symbol at runtime.
-/// \p size is the size in bytes of the symbol.
-/// \p offset is the offset in bytes from the base address.
+
```
Comment (opti-mix) [inline include/glow/Backends/CompiledFunction.h]: No empty lines between the doxygen comments and the code.

File: include/glow/Backends/CompiledFunction.h [hunk 31]
```diff
@@ -16,12 +16,47 @@
...
 #define GLOW_BACKENDS_COMPILEDFUNCTION_H
 
+#include "glow/Graph/Nodes.h"
 #include <unordered_map>
 
 namespace glow {
 
 class Context;
+namespace runtime {
+/// Runtime Symbol Info
```
Comment (qcolombet) [inline include/glow/Backends/CompiledFunction.h]: Nit: I fell this should be the class comment and need to be more explanatory.
http://llvm.org/docs/CodingStandards.html#commenting

File: include/glow/Backends/CompiledFunction.h [hunk 32]
```diff
@@ -16,12 +16,47 @@
...
+/// Contains Information for initialization and handling of symbol at runtime.
 
+struct RuntimeSymbolInfo {
+  /// \p size is the size in bytes of the symbol.
+  size_t size;
+
+  /// \p offset is the offset in bytes from the base address.
+  size_t offset;
```
Comment (qcolombet) [inline include/glow/Backends/CompiledFunction.h]: More a remark than anything. With an offset we assume that the runtime has easily access to the base address and that it knows easily how to map a symbol to a base address.
For instance, let say I ask where is `foo`, we need to know somewhere that `foo` is a weight or activation to look at the right table.
More generally, again just future looking remarks, if we start to share weights between different graphs (as in not different instances of the same graph, but really different graphs[1]), we may want to create more fine tuned pool of symbols and we would need to somehow keep track of which symbol belongs to which pool.
The bottom line is moving forward, I wonder if this should be a full address instead of just an offset (or a base + offset in case we want to support relocating things)

[1] This could easily happen if we were to split a graph in the middle. E.g.,
Split
```
Weights
  |   \
  V    \
 op     \
  |     |
  V     |
 cpu Op |
  \     |
   \   /
    V V
    op
```
Into
```
Weights   [shared between G1 & G3]
--|---\---- vv G1 on accelerator A
  V    \
 op     \
--|-----|--- vv G2 on CPU
  V     |
 cpu Op |
  \     |
---\---/--- vv G3 on accelerator A
    V V
    op
```

File: include/glow/Backends/CompiledFunction.h [hunk 33]
```diff
@@ -16,23 +16,50 @@
...
+  /// The size in bytes.
+  size_t size;
+  /// Offset in bytes from the base address.
+  size_t offset;
+};
+/// Runtime Bundle
+/// Contains the information needed to be passed forward from compile time to
+/// runtime. In order to allocate and initialize memory.
+struct RuntimeBundle {
+  /// Map from symbol name to a RuntimeSymbolObject.
```
Comment (qcolombet) [inline include/glow/Backends/CompiledFunction.h]: RuntimeSymbolObject -> RuntimeSymbolInfo

File: include/glow/Backends/CompiledFunction.h [hunk 34]
```diff
@@ -16,23 +16,50 @@
...
+  std::unordered_map<std::string, RuntimeSymbolInfo> symbolTable;
+  /// Pointer to memory containing the weights for execution.
+  uint8_t *constants;
+  /// Amount of memory needed for weights.
+  const size_t constantWeightVarsMemSize;
+  /// Mmount of memory needed for mutable vars.
```
Comment (qcolombet) [inline include/glow/Backends/CompiledFunction.h]: Mmount -> Amount

File: include/glow/Backends/CompiledFunction.h [hunk 35]
```diff
@@ -16,23 +16,50 @@
...
+  const size_t mutableWeightVarsMemSize;
+  /// Amount of memory needed for activations.
+  const size_t activationsMemSize;
```
Comment (qcolombet) [inline include/glow/Backends/CompiledFunction.h]: Just checking, what is the reason we split mutableWeight and activations? Is to do something special for training?
Otherwise and I don't really see why we would need to track them differently.

File: include/glow/Backends/CompiledFunction.h [hunk 36]
```diff
@@ -16,23 +16,50 @@
...
+  RuntimeBundle(size_t constWeight, size_t mutableWeight, size_t activations)
+      : constantWeightVarsMemSize(constWeight),
+        mutableWeightVarsMemSize(mutableWeight),
+        activationsMemSize(activations) {}
```
Comment (qcolombet) [inline include/glow/Backends/CompiledFunction.h]: It feels weird that this API doesn't have anything to manage the symbolTable and the constants array.
It feels weird as well to have the size for the various section but the pointer only for the constants. Could you elaborate on how to you see this being used?

File: lib/Backends/CPU/AllocationsInfo.h [hunk 37]
```diff
@@ -16,6 +16,8 @@
 #ifndef GLOW_BACKENDS_CPU_ALLOCATIONSINFO_H
 #define GLOW_BACKENDS_CPU_ALLOCATIONSINFO_H
 
+// #include "glow/Backends/CompiledFunction.h"
```
Comment (qcolombet) [inline lib/Backends/CPU/AllocationsInfo.h]: Unused code ;)

File: lib/Backends/CPU/CPUBackend.h [hunk 38]
```diff
@@ -47,6 +47,9 @@ class CPUBackend : public BackendUsingGlowIR {
 
   std::unique_ptr<CompiledFunction> compile(Function *F,
                                             const Context &ctx) const override;
+  /// Loads the function binary and constants onto device.
+  /// For CPU case this is primarily just allocating constant memory.
+  void loadFunctionOnDevice(const Context &ctx);
```
Comment (qcolombet) [inline lib/Backends/CPU/CPUBackend.h]: I believe this API should be in a different class. Like DeviceManager that each backend needs to implement.
Basically, a backend should come with different and separate APIs, at least two, compile and runtime.

File: include/glow/Backends/CompiledFunction.h (hunk 36 above)
Comment (gcatron) [reply include/glow/Backends/CompiledFunction.h]: These fields could be passed to the CompiledFunction directly but a struct gives us flexibility going forward to support different backend needs.

File: lib/Backends/CPU/CPUBackend.h (hunk 38 above)
Comment (gcatron) [reply lib/Backends/CPU/CPUBackend.h]: So that particular function was meant to be deleted :P But! I agree I created #1951 to track designing and creating an API for runtime.

File: include/glow/Backends/CompiledFunction.h (hunk 36 above)
Comment (qcolombet) [reply include/glow/Backends/CompiledFunction.h]: What I am trying to get at is from a design perspective, one could mess with the constants pointer without doing the proper changes to the weight size member. That's why I am surprised it is not encapsulate behind some APIs. I.e., we expect the users of this API to do the right thing and that's when I get suspicious :).

Now, from a more high level point of view, the handling of constants and mutable allocations are the same. The only difference is that we know the contain of the constants at compile time. Therefore what I am questioning is whether or not this special handling of constants makes sense.

File: include/glow/Backends/CompiledFunction.h (hunk 32 above)
Comment (gcatron) [reply include/glow/Backends/CompiledFunction.h]: Summarizing the conversations between @opti-mix  and @qcolombet. I think adding a "kind" value to symbolInfo that allows interchange between offset and absolute memory addressing is a great idea, but not needed for this PR. We will revisit when we looks at expanding the runtime API perhaps in #1951 .

File: include/glow/Backends/CompiledFunction.h (hunk 35 above)
Comment (gcatron) [reply include/glow/Backends/CompiledFunction.h]: Due to the difference in the observable nature of Mutable vs activations. Mutable is user observable and may be placed in a different region of memory compared to activations.

Comment (qcolombet) [reply include/glow/Backends/CompiledFunction.h]: Sounds good. We should clarify this in the comments.

File: lib/Backends/CPU/CPUFunction.cpp [hunk 39]
```diff
@@ -13,28 +13,80 @@
...
+#include "glow/Graph/Context.h"
 #include "glow/Support/Compiler.h"
 #include "glow/Support/Memory.h"
 
 using namespace glow;
 
-CPUFunction::CPUFunction(std::unique_ptr<llvm::orc::GlowJIT> JIT, void *heap)
-    : JIT_(std::move(JIT)), heap_(heap) {}
+CPUFunction::CPUFunction(std::unique_ptr<llvm::orc::GlowJIT> JIT,
+                         runtime::RuntimeBundle runtimeBundle)
```
Comment (rdzhabarov) [inline lib/Backends/CPU/CPUFunction.cpp]: you can pass by const reference here.

File: lib/Backends/CPU/CPUFunction.cpp [hunk 40]
```diff
@@ -13,28 +13,80 @@
...
+  copyMutablesToDevice(ctx);
+
   auto sym = JIT_->findSymbol("jitmain");
   assert(sym && "Unable to JIT the code!");
-  using JitFuncType = void (*)(void);
+  using JitFuncType =
+      void (*)(uint8_t * constantWeightVars, uint8_t * mutableWeightVars,
+               uint8_t * activations);
   auto address = sym.getAddress();
   if (address) {
```
Comment (rdzhabarov) [inline lib/Backends/CPU/CPUFunction.cpp]: I'm curious what is the situation when sym.getAddress() is null? Could it ever happen?

File: lib/Backends/CPU/CPUFunction.cpp [hunk 41]
```diff
@@ -13,28 +13,80 @@
...
+    auto payload = baseMutableWeightVarsAddress_ + symbolInfo->second.offset;
+    auto numBytes = symbolInfo->second.size;
+    auto addr = PH.second->getUnsafePtr();
+    // copy PH from allocated memory.
+    memcpy(addr, payload, numBytes);
+  }
+}
+
+void CPUFunction::freeAllocations() {
+  free(baseMutableWeightVarsAddress_);
```
Comment (rdzhabarov) [inline lib/Backends/CPU/CPUFunction.cpp]: must be alignedFree! :-)

File: lib/Backends/CPU/CPUFunction.cpp [hunk 42]
```diff
@@ -13,28 +13,80 @@
...
+  free(baseActivationsAddress_);
```
Comment (rdzhabarov) [inline lib/Backends/CPU/CPUFunction.cpp]: ditto

**Comment by rdzhabarov** [comment]
Have you verified that AOT still works :)?

File: lib/Backends/CPU/CPUFunction.cpp (hunk 40 above)
Comment (gcatron) [reply lib/Backends/CPU/CPUFunction.cpp]: > Have you verified that AOT still works :)?

Testing now, it has one core at 100% and has for the last 5 minutes....

File: lib/Backends/CPU/CPUBackend.cpp [hunk 43]
```diff
@@ -41,16 +41,17 @@ namespace {
...
-/// Emit the entry point for JIT called "jitmain". It simply calls the main
-/// entry of the module with the constant concrete addresses of all the memory
-/// areas. Since these addresses are constants, the LLVM optimizer will constant
-/// propagate them into relative addressing computations and the like and
-/// produce a very efficient code that uses absolute addressing whenever
-/// possible.
+/// Emit the entry point for JIT called "jitmain".
+/// Function has the following API:
+///   void jitmain(uint8_t *baseConstantWeightVars, uint8_t
+///   *baseInoutWeightVars,
```
Comment (rdzhabarov) [inline lib/Backends/CPU/CPUBackend.cpp]: nit: baseIn"Out"

File: lib/Backends/CPU/CPUFunction.cpp [hunk 44]
```diff
@@ -13,28 +13,80 @@
...
+  baseActivationsAddress_ = (uint8_t *)alignedAlloc(
+      runtimeBundle_.activationsMemSize, TensorAlignment);
+
+  baseMutableWeightVarsAddress_ = (uint8_t *)alignedAlloc(
+      runtimeBundle_.mutableWeightVarsMemSize, TensorAlignment);
+}
+
+void CPUFunction::copyMutablesToDevice(Context &ctx) {
+  // Copy Placeholders into allocated memory.
+  for (auto PH : ctx.pairs()) {
```
Comment (rdzhabarov) [inline lib/Backends/CPU/CPUFunction.cpp]: placeholders contain both inputs and outputs. Do we need to copy outputs? What's your plan here?

File: lib/Backends/CPU/CPUFunction.h [hunk 45]
```diff
@@ -27,21 +27,35 @@ class CPUFunction final : public CompiledFunction {
...
-
+  CPUFunction(std::unique_ptr<llvm::orc::GlowJIT> JIT,
+              const runtime::RuntimeBundle &runtimeBundle);
+  /// Copy Function to device, an empty function for CPU.
+  void copyFunctionToDevice(){};
+  /// Copy Constants to device, an empty function for CPU.
+  void copyConstantsToDevice(){};
+  /// Allocate Mutable buffers on device, Activations and Placeholders.
+  void allocateMutableBuffersOnDevice();
+  /// Copy Mutable data to device.
```
Comment (rdzhabarov) [inline lib/Backends/CPU/CPUFunction.h]: what's actually "mutable data"?

It's clear what's constants, activations, placeholders.

Comment (gcatron) [reply lib/Backends/CPU/CPUFunction.h]: I don't know if we currently have a good way to filter this from the context. For this PR I suggest we copy all. I've created a separate issue #1953 to track how we'd like to pass inputs/outputs going forward.

File: lib/Backends/CPU/CPUFunction.cpp (hunk 44 above)
Comment (gcatron) [reply lib/Backends/CPU/CPUFunction.cpp]: I don't know if we currently have a good way to filter this from the context. For this PR I suggest we copy all. I've created a separate issue #1953 to track how we'd like to pass inputs/outputs going forward.

File: lib/Backends/CPU/CPUFunction.cpp (hunk 40 above)
Comment (opti-mix) [reply lib/Backends/CPU/CPUFunction.cpp]: It is very unlikely to happen, but we can add an assert here just for the case.

**Comment by nadavrot** [comment]
Congrats! Nice work.
